### ISSUE TITLE
Memory leak when using fetch requests in react native

### ISSUE THREAD

**Issue by Vadko** [facebook/hermes#1147]
## Bug Description

https://github.com/facebook/react-native/issues/39100

Issue above is opened on react-native repo but after investigation I made I can confirm that it is related definitely to hermes. On react-native 0.69 there is no memory leak when hermes is turned off.

I've rechecked it on latest version and behaviour is the same - memory leak with turned on hermes, proper ram cleaning when hermes is turned on. Provided demo in the issue above is for the newer react native version.

- [x] I have run `gradle clean` and confirmed this bug does not occur with JSC

Hermes version: 0.11
React Native version (if any): 0.69, 0.72.5
OS version (if any):
Platform (most likely one of arm64-v8a, armeabi-v7a, x86, x86_64):

## Steps To Reproduce

1.
2.

code example:
https://github.com/clemensmol/rn-fetch-memoryleak
## The Expected Behavior
No memory leak when doing fetch requests

**Comment by tmikov** [comment]
There are differences in native integration between Hermes and JSC, differences in GC behavior, etc, so it is impossible to say with certainty that this is a bug **in** Hermes, just because it is not readily observed in JSC.

Of course we also **cannot** exclude the possibility that this is a bug in Hermes, but Hermes does not have builtin functionality for fetching http requests. So, presumably this is implemented by a non-trivial amount of C++ in React Native? 

It is not practical for the Hermes team to take ownership of large unrelated library code to find a bug which may or may not be caused by Hermes. We will not ignore this, but we will likely need help in creating a very narrow repro that points definitely to Hermes.

As a first step, @Vadko, could you try running `globalThis.gc()` periodically, or perhaps after each fetch request completes, and see if it affects anything?

**Comment by domeknn** [comment]
@tmikov as far as I know, team of React Native didnt change anything related to Networking for very long time. I tried to call `globalThis.gc()` but it doesn't freeing up that much


https://github.com/facebook/hermes/assets/9402280/ebd61c73-0c3f-404f-9354-07f3c58fab87

**Comment by Vadko** [comment]
@tmikov confirming what @domeknn reported, works same in my case as well

**Comment by tmikov** [comment]
Got it. We will try to reproduce it internally. Thanks for creating the repro!

**Comment by neildhar** [comment]
Hey @Vadko, thanks for preparing this repro. I have been trying to reproduce the issue, and do see a steep sustained increase in memory consumption without the call to `globalThis.gc()` on both Android and iOS. (I modified your repro to invoke fetch every 50ms instead of 500ms to make the issue more severe)

I then modified your example to periodically invoke the GC (every 100 ms). With that change, the Android memory increase seemed to be resolved. On iOS, I still observed a very gradual leak. Looking at it under XCode's allocation profiler, the primary culprit is the `CFString (immutable)` class of allocations although I also see the `NSURLInternal` size increasing over time. I don't know what the source of these allocations are.

However, that gradual leak also appears under JSC, which means that isn't a Hermes specific issue.

So to summarise some inferences:
1. Hermes is not running the GC frequently enough, which is causing network requests to accumulate and take up memory. This is a known difference from JSC, that (for now) can be addressed by manually calling `globalThis.gc()`.
2. There seems to be a real memory leak somewhere in the networking code for React Native. The leak is clearly present in the repro you shared in the two allocation classes I identified above, and appears under both Hermes and JSC. Someone familiar with React Native should investigate the source of these allocations.

**Comment by domeknn** [comment]
> Hey @Vadko, thanks for preparing this repro. I have been trying to reproduce the issue, and do see a steep sustained increase in memory consumption without the call to `globalThis.gc()` on both Android and iOS. (I modified your repro to invoke fetch every 50ms instead of 500ms to make the issue more severe)
> 
> I then modified your example to periodically invoke the GC (every 100 ms). With that change, the Android memory increase seemed to be resolved. On iOS, I still observed a very gradual leak. Looking at it under XCode's allocation profiler, the primary culprit is the `CFString (immutable)` class of allocations although I also see the `NSURLInternal` size increasing over time. I don't know what the source of these allocations are.
> 
> However, that gradual leak also appears under JSC, which means that isn't a Hermes specific issue.
> 
> So to summarise some inferences:
> 
>     1. Hermes is not running the GC frequently enough, which is causing network requests to accumulate and take up memory. This is a known difference from JSC, that (for now) can be addressed by manually calling `globalThis.gc()`.
> 
>     2. There seems to be a real memory leak somewhere in the networking code for React Native. The leak is clearly present in the repro you shared in the two allocation classes I identified above, and appears under both Hermes and JSC. Someone familiar with React Native should investigate the source of these allocations.

About 1. :
In my case hermes never calls gc so it accumulate all of the fetches :( I need to run it manually to flush the memory a bit. Maybe it's worth to check why it doesnt call? 🤔

**Comment by neildhar** [comment]
> Maybe it's worth to check why it doesnt call

The problem here is that the memory allocated for the fetched data is not visible to the GC, since it was allocated by native code. We've considered adding a mechanism to inform the GC about these allocations, but that would still require making changes to the allocating code to report them.

> also on dev build gc doesnt work well in this case

My testing was done on dev builds, and I did not observe significant sustained increases in memory consumption with Hermes. Some modest initial increase in memory consumption is expected as you start actually running the code, but it should even out as you run 10s or 100s of requests. Have you tried this?

**Comment by Almouro** [comment]
Hi!

We're facing a similar issue in one of our apps with low memory usage constraints, and can indeed reproduce with the example that @vadko provided 👍 

I've changed the example to make more API calls ([full code here](https://gist.github.com/Almouro/582c867fec58dc974ee2107a2f0083c7))

Essentially, I:

- **start the app**
  - RAM is ~130MB
- **start making 100s of API calls**
  - RAM goes up to ~400MB (or higher if we don't stop)
- **stop making API calls**
  - RAM stagnates at ~400MB
- **trigger manually JS GC**
  - RAM goes down again to ~130MB

Im still having a hard time understanding why Hermes isn't garbage collecting as expected. Garbage collection does seem to be happening as shown in the details below, but is not fully collecting everything.

The fact that running `gc` manually fixes the RAM increase makes it seem to me that there's no memory leak, and perhaps no issue on React Native `fetch` side?

Plus, as shown in the details below, in debug, there seems to be a discrepancy by the JS size reported by the Hermes debugger (which increases as expected) and the JS size reported by `HermesInternal.getInstrumentedStats()` (which doesn't)

Would love some help investigating further though 🙏
It also seems linked to https://github.com/facebook/hermes/issues/982

## Experiment details

### Device

Measures shown below are taken with a Pixel 8 Android 14, however we've also tested with Hades incremental on a Android 9 device

Here's the result of `HermesInternal.getRuntimeProperties()` for the Pixel:

```
{
  'Bytecode Version': 94,
  'Builtins Frozen': false,
  'VM Experiments': 0,
  Build: 'Release',
  GC: 'hades (concurrent)',
  'OSS Release Version': 'for RN 0.72.6',
  'CommonJS Modules': 'None'
}
```

### 1. First, let's start the app

Metrics gathered from `adb shell dumpsys meminfo` and `HermesInternal.getInstrumentedStats()`

| Metric                          | Value |
| ------------------------------- | ----- |
| RAM used:                       | 126MB |
| Java Heap                       | 3MB   |
| Native Heap                     | 10MB  |
| Hermes JS allocated bytes       | 2MB   |
| Hermes Total JS allocated bytes | 2MB   |

This is the full result of `HermesInternal.getInstrumentedStats()`

```
{ js_VMExperiments: 0,
  js_numGCs: 0,
  js_gcCPUTime: 0,
  js_gcTime: 0,
  js_totalAllocatedBytes: 1759136,
  js_allocatedBytes: 1759136,
  js_heapSize: 8388608,
  js_mallocSizeEstimate: 0,
  js_vaSize: 8388608,
  js_markStackOverflows: 0 }
```

### 2. Then let's make a few 100s API calls and stop after a while

RAM rises up progressively. Note that it would go up until crashing the app or receiving a memory pressure event and finally running Hades Garbage collection.

| Metric                          | Value |
| ------------------------------- | ----- |
| RAM used:                       | 429MB |
| Java Heap                       | 212MB |
| Native Heap                     | 103MB |
| Hermes JS allocated bytes       | 5MB   |
| Hermes Total JS allocated bytes | 214MB |

```
{ js_VMExperiments: 0,
  js_numGCs: 112,
  js_gcCPUTime: 0.3280000000000002,
  js_gcTime: 0.3430000000000002,
  js_totalAllocatedBytes: 213710856,
  js_allocatedBytes: 5517344,
  js_heapSize: 12582912,
  js_mallocSizeEstimate: 0,
  js_vaSize: 12582912,
  js_markStackOverflows: 0 }
```

The number of GC triggered also went up on JS side, which is surprising! Garbage collection does seems to happen which explains why js allocated bytes stays low. But is the stat correct? As shown at the end of the report, there's a discrepancy with the one reported in the Hermes debugger for instance

On Java side, the garbage collector is also trying to free memory but can't (and increases heap size), probably because the JS side holds the instances.

### 3. Wait a few minutes

In case garbage collection happens, but nope, no changes
The number of GC on JS side is the same.

### 4. Trigger manually JS garbage collection

| Metric                          | Value |
| ------------------------------- | ----- |
| RAM used:                       | 133MB |
| Java Heap                       | 4MB   |
| Native Heap                     | 19MB  |
| Hermes JS allocated bytes       | 1MB   |
| Hermes Total JS allocated bytes | 214MB |

```
{ js_VMExperiments: 0,
  js_numGCs: 116,
  js_gcCPUTime: 0.4140000000000002,
  js_gcTime: 0.4300000000000002,
  js_totalAllocatedBytes: 214355544,
  js_allocatedBytes: 958096,
  js_heapSize: 8388608,
  js_mallocSizeEstimate: 0,
  js_vaSize: 8388608,
  js_markStackOverflows: 0 }
```

## Running in debug mode

Making the same experiment in debug and checking out the Hermes debugger in Flipper, we can see the heap snapshot increases in size, while `HermesInternal.getInstrumentedStats()` doesn't seem to report it.

Biggest difference between my 2 heap snapshots are the 78k strings allocated, corresponding to the request payload:

![image](https://github.com/facebook/hermes/assets/4534323/7bb80c74-101c-4d7f-8cab-b88aaf096b0c)

- Stats after starting the app:

RAM used: 217MB

```
{ js_VMExperiments: 0,
  js_numGCs: 1,
  js_gcCPUTime: 0.002,
  js_gcTime: 0.002,
  js_totalAllocatedBytes: 6121376,
  js_allocatedBytes: 3227552,
  js_heapSize: 8388608,
  js_mallocSizeEstimate: 0,
  js_vaSize: 8388608,
  js_markStackOverflows: 0 }
```

- Stats after making 100s of API calls and stop

RAM used 490MB

```
{ js_VMExperiments: 0,
  js_numGCs: 173,
  js_gcCPUTime: 2.457999999999995,
  js_gcTime: 2.573999999999996,
  js_totalAllocatedBytes: 220705392,
  js_allocatedBytes: 9991808,
  js_heapSize: 20971520,
  js_mallocSizeEstimate: 0,
  js_vaSize: 20971520,
  js_markStackOverflows: 0 }
```

**Comment by tmikov** [comment]
> Im still having a hard time understanding why Hermes isn't garbage collecting as expected.

@Almouro Hermes is actually garbage collecting exactly as expected. A garbage collector is not responsible for memory (or other resources like file handles) it did not allocate itself. This is true and has always been true in garbage collected languages like Java or .NET. In fact the garbage collector is not required to even run at all, if the allocated memory is under a given threshold. That is why the recommendation is to **always explicitly free native resources**. If you open a file, you need to explicitly close it. Similarly, if you allocate native memory, you need to explicitly free it and not rely on the GC.

With that said, we understand that this is a real problem that people are facing, so we will be adding a method to "inform" the GC of native memory that has been allocated. However, library writers will have to call that method.

**Comment by Almouro** [comment]
Thanks for your answer @tmikov  🙏
Btw, the Hades are very detailed which is great! so if I've missed a detail that is mentioned there, I do apologize 🙏

Indeed, I probably should have phrased that "as I expected" instead of "as expected" 😅, as there are some things that I don't understand yet, especially:

1. Why does RAM continuously increase and not stagnate at some point?

My understanding (which is probably wrong or erroneous) is that JS objects hold Java and native objects, especially since triggering manually the JS GC clears all memory, native and Java included.
When JS heap of objects reaches a certain size (I imagine it should be the JS heap size in the Hermes internal), I imagine garbage collecting happens (which seems to work properly since we can see the number of `gc` increasing in the Hermes internals)

If I follow this reasoning (again probably erroneous), it seems to me that at some point, RAM should stagnate after reaching a certain point.

Say we have 1MB of JS holding 10MB in Native/Java. If our JS heap size increases to ~10MB, that would mean the RAM should reach ~10MB of JS + 100MB of Native/Java, and then Hades would start collecting to free up memory (or increase the heap size). So the app would not use more 110MB of RAM and stagnate around that point.

Instead in our case, it keeps rising until the OS sends a memory pressure alert (which triggers a JS GC inside React Native)

2. Why is Hermes debugger reporting 110MB for the JS memory snapshot when the heap size is 8MB?

This is way above the JS heap size reported by the Hermes internal. Is this because this size in the Hermes debugger also includes native allocations (java & cpp)?
Knowing the reason would in debugging those kind of issues with external libraries.


Any help in understanding is really appreciated! 🙏

**Comment by neildhar** [comment]
@Almouro In principle, your reasoning is correct. The problem is that the steady state you are describing may be much larger than the available memory on the system. For instance, in this case, it seems like each object (~48 bytes) retains 250KB of native memory. So just ~1MB of JS memory can retain ~5GB of native memory. Because of this, Hermes does not end up collecting sufficiently frequently to keep the heap size low.

> this size in the Hermes debugger also includes native allocations

This is an interesting observation. There is a subset of off-heap memory that *is* managed by the GC. This is used to back strings and `ArrayBuffer`s. These seem to account for the difference in consumption reported by `getInstrumentedStats` and your heap snapshot. It would be good to check if these might be the culprit. Some questions that would be useful to answer that:

1. Does the size of the snapshot continue to increase over time? Does that size increase account for the increase in overall memory consumption?
2. Does the size of the snapshot shrink when you have regular calls to `gc()`?

**Comment by Almouro** [comment]
Thanks @neildhar for your help!

1. The size of the snapshot continues to increase over time, but doesn't account for the whole memory consumption
2. The size of the snapshot shrinks back to the initial state after a call to `gc()`

## Experiments with Hermes debugger

Here are some results in debug mode with some Hermes debugger snapshots, where I:
- start the app
- make 100s of API calls
- make more API calls
- trigger GC manually

First 3 stats are taken from `dumpsys meminfo`, last 4 from `getInstrumentedStats()`

| Metric                                | Initial State | 100s of API calls | More API calls | After gc       |
| ------------------------------------- | ------------- | ----------------- | -------------- | -------------- |
| **Total RAM**                             | 210MB         | 483MB **(+273MB)**    | 716MB **(+233MB)** | 262MB **(-454MB)** |
| **Java Heap**                             | 11MB          | 95MB **(+84MB)**      | 197MB **(+102MB)** | 10MB **(-187MB)**  |
| **Native Heap**                           | 65MB          | 237MB **(+172MB)**    | 411MB **(+174MB)** | 51MB **(-360MB)**  |
| **Flipper Hermes Debugger snapshot size** | 16.2MB        | 113MB **(+97MB)**     | 296MB **(+183MB)** | 17MB **(-279MB)**  |
| **Hermes JS allocated bytes**             | 2.5MB         | 13.9MB **(+11MB)**    | 53MB **(+39MB)**   | 2MB **(-51MB)**    |
| **Hermes Total JS allocated bytes**       | 5.4MB         | 218MB    | 576MB  | 577MB   |
| **JS Heap size**                          | 8.4MB         | 25MB              | 59MB           | 55MB           |
| **JS num GC**                             | 1             | 242               | 611            | 615            |

## Size of objects

> For instance, in this case, it seems like each object (~48 bytes) retains 250KB of native memory. So just ~1MB of JS memory can retain ~5GB of native memory

Right, I didn't consider that it could be that bad a ratio 😅 Since JS garbage collecting does happen even when not triggered manually, I thought the ratio was smaller. 
How did you find that it was 48 bytes of JS vs the 250KB of native in this example?

I also hadn't realized the JS heap size was increasing (8MB to 59MB), can it increase as much as it can until it reaches the max heap size?

In React Native, seems like we initialize Hermes somewhere around here https://github.com/facebook/react-native/blob/9a3b75c7824e1b8896ffc4f05c3386587513f2be/packages/react-native/ReactCommon/hermes/executor/HermesExecutorFactory.cpp#L233

With 
```cpp
  return ::hermes::vm::RuntimeConfig::Builder()
      .withEnableSampleProfiling(true)
      .build();
  ```
But adding a breakpoint here, it seems the default max heap size is 3GB:
<img width="491" alt="image" src="https://github.com/facebook/hermes/assets/4534323/a351f570-10b0-490d-8ac2-5c7eb453987f">


Happy to provide more data if needed! 🙏

**Comment by neildhar** [comment]
@Almouro Thank you for sharing these detailed findings. They exhibit very unusual memory behaviour, and were exactly the evidence I needed for a full investigation of what is going on.

In particular, the heap snapshots showing that Hermes is allocating and retaining very large strings that are only freed by explicit calls to the GC suggested that something was affecting the GC's ability to collect the large strings.

I've spent some time investigating this and I believe the root cause of the behaviour you're observing is a very subtle interaction between `WeakMap`s and the GC. I have a relatively simple mitigation that addresses the collection of those strings, and improves memory consumption dramatically in your repro.

That said, this isn't a general solution for the problem in this issue. The underlying issue with untracked native resources remains, and the repro in the initial report in this issue still stands.

> How did you find that it was 48 bytes of JS vs the 250KB of native in this example

Objects in Hermes are around 48 bytes, and the 250KB was based on the size of the blob being downloaded. This is just a rough estimate, for the purpose of discussion.

> can it increase as much as it can until it reaches the max heap size?

It certainly can, but only if the application needs a lot of working memory. In this case, the GC was overestimating the amount of working memory the application needed because it couldn't collect the large strings.

**Comment by JunielKatarn** [comment]
Also reproducible here:
https://github.com/jurocha-ms/Repro/tree/main/ReactNative/fetchwinleak

**Comment by neildhar** [comment]
Closing this since the WeakMap leak has been fixed, and we have added an API for tracking external memory in aae2c4260781178d7b2ca169811b3bfca9f924d2. That allows you to inform the GC that a given object retains some native memory.
